### PR TITLE
Static comparators

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt [hunk 1]
```diff
@@ -181,11 +181,15 @@ internal class ListLicensesCommand : CliktCommand(
 private data class TextLocationGroup(
     val locations: Set<TextLocation>,
     val text: String? = null
-)
+) {
```
Comment (fviernau) [inline helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: Commit-message: What performance problem do we actually have and how much does it help?

Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: > What performance problem do we actually have and how much does it help?

I've improved the commit message in that regard. I did not benchmark the changes, though, as I believe they are beneficial even when leaving the performance aspect aside.

> _edit_: I'm asking because I find the code gets less readable in some places after this change which I know is subjective.

I only saw this edit after I've reworded the commit message. Looks like we again have different 
views, as I find the code to now be more readable throughout the changes.

Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: > I've improved the commit message in that regard. I did not benchmark the changes, though, as I believe they are beneficial even when leaving the performance aspect aside.

Thanks, this addressed actually my actually nit concerns. 

> I only saw this edit after I've reworded the commit message. Looks like we again have different views, as I find the code to now be more readable throughout the changes.

Sorry, I've mixed things up a bit with the other commit in this PR. I actually was referring to the readability change when using this type alias.

Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: > I actually was referring to the readability change when using this type alias.

Ah, ok. I agree introducing typealiases (only) for Path- and ScopeExcludes, and only for non-mutable but not for mutable lists, is a *bit* arbitrary. I initially made that change as some lines otherwise would have exceeded 120 chars, and they couldn't be wrapped nicely. But meanwhile I've changed the code anyway and I could drop the typealias change if you prefer.

Comment (mnonnenmacher) [reply helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: I would prefer to drop that commit, we use so many lists throughout the code that it's really inconsistent to add typealiases for two of the generic types, and I agree that it makes it harder to read the code, because you need to know the typealias.

Comment (mnonnenmacher) [inline helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: Commit message nit: Maybe write out "especially"?

File: model/src/main/kotlin/TextLocation.kt [hunk 2]
```diff
@@ -42,10 +42,12 @@ data class TextLocation(
...
     companion object {
-        private val COMPARATOR = compareBy<TextLocation>({ it.path }, { it.startLine }, { it.endLine })
-        val SORTED_SET_COMPARATOR = SortedSetComparator<TextLocation>()
-        val TREE_SET_TYPE by lazy { jsonMapper.typeFactory.constructTreeSetType(TextLocation::class.java) }
         const val UNKNOWN_LINE = -1
+
+        val TREE_SET_TYPE by lazy { jsonMapper.typeFactory.constructTreeSetType(TextLocation::class.java) }
+
+        val SORTED_SET_COMPARATOR = SortedSetComparator<TextLocation>()
+        private val COMPARATOR = compareBy<TextLocation>({ it.path }, { it.startLine }, { it.endLine })
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/TextLocation.kt:50]: Nit: This change should be in a separate commit.

File: helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt (hunk 1 above)
Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: > I would prefer to drop that commit

Done, although ...

> because you need to know the typealias.

... this would mean we should *never* use typealiases, which I don't agree to.

File: reporter/src/main/kotlin/Reporter.kt [hunk 3]
```diff
@@ -20,11 +20,22 @@
...
 import java.util.ServiceLoader
+import java.util.SortedMap
+
+internal val SCOPE_EXCLUDE_LIST_COMPARATOR = compareBy<Map.Entry<String, ScopeExcludes>>(
+    { it.value.isNotEmpty() }, { it.key }
+)
+
+internal val SCOPE_EXCLUDE_MAP_COMPARATOR = compareBy<Map.Entry<Identifier, SortedMap<String, ScopeExcludes>>>(
+    { it.value.values.isNotEmpty() }, { it.key }
+)
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/Reporter.kt:38]: I'd prefer to put this in a `Comparators.kt` file in the utils package.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 4]
```diff
@@ -45,6 +45,11 @@ internal val VCS_DIRECTORIES = listOf(
     "CVS"
 )
 
+/**
+ * A comparator that sorts parents paths before child paths.
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxUtils.kt:49]: "parent~~s~~ paths"

File: helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt (hunk 1 above)
Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/ListLicensesCommand.kt:184]: Done.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt (hunk 4 above)
Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt:49]: Done.

File: model/src/main/kotlin/TextLocation.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/main/kotlin/TextLocation.kt:50]: Done.

File: reporter/src/main/kotlin/Reporter.kt (hunk 3 above)
Comment (sschuberth) [reply reporter/src/main/kotlin/Reporter.kt]: Done.
